test/Microsoft.AspNetCore.Http.Connections.Tests/HttpConnectionDispatcherTests.cs

```diff
@@ -1797,6 +1797,137 @@ bool ExpectedErrors(WriteContext writeContext)
             }
         }
 
+        private class ControllableMemoryStream : MemoryStream
+        {
+            private readonly TaskCompletionSource<bool> _tcs;
+
+            public ControllableMemoryStream(TaskCompletionSource<bool> tcs)
+            {
+                _tcs = tcs;
+            }
+
+            public override async Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                await _tcs.Task;
+
+                await base.CopyToAsync(destination, bufferSize, cancellationToken);
+            }
+        }
+
+        [Fact]
+        public async Task WriteThatIsDisposedBeforeCompleteReturns404()
+        {
+            using (StartVerifiableLog(out var loggerFactory, LogLevel.Debug))
+            {
+                var manager = CreateConnectionManager(loggerFactory);
+                var pipeOptions = new PipeOptions(pauseWriterThreshold: 13, resumeWriterThreshold: 10);
+                var connection = manager.CreateConnection(pipeOptions, pipeOptions);
+                connection.TransportType = HttpTransportType.LongPolling;
+
+                var dispatcher = new HttpConnectionDispatcher(manager, loggerFactory);
+
+                var services = new ServiceCollection();
+                services.AddSingleton<TestConnectionHandler>();
+                var builder = new ConnectionBuilder(services.BuildServiceProvider());
+                builder.UseConnectionHandler<TestConnectionHandler>();
+                var app = builder.Build();
+                var options = new HttpConnectionDispatcherOptions();
+
+                TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();
+
+                using (var responseBody = new MemoryStream())
+                using (var requestBody = new ControllableMemoryStream(tcs))
+                {
+                    var context = new DefaultHttpContext();
+                    context.Request.Body = requestBody;
+                    context.Response.Body = responseBody;
+                    context.Request.Path = "/foo";
+                    context.Request.Method = "POST";
+                    var values = new Dictionary<string, StringValues>();
+                    values["id"] = connection.ConnectionId;
+                    var qs = new QueryCollection(values);
+                    context.Request.Query = qs;
+                    var buffer = Encoding.UTF8.GetBytes("Hello, world");
+                    requestBody.Write(buffer, 0, buffer.Length);
+                    requestBody.Seek(0, SeekOrigin.Begin);
+
+                    // Write. This will wait on the TCS inside ApplicationStream.CopyToAsync
+                    var sendTask1 = dispatcher.ExecuteAsync(context, options, app);
+
+                    // Extra check to make sure ExecuteAsync is waiting on the TCS
+                    if (sendTask1.Status != TaskStatus.WaitingForActivation)
```

Comment:
FYI this is what `SyncPoint` in the `Tests.Utils` assembly is for. It's a pair of TCSes that let you do basically the same thing.

You would give the SyncPoint to the `ControllableMemoryStream` and have it call and `await` `.WaitToContinue()` in `CopyToAsync`. Then outside the code, after calling `ExecuteAsync` you await `.WaitForSyncPoint()`.

`WaitForSyncPoint` is released when `WaitToContinue` is called. Then `Continue` releases `WaitToContinue`.

Reply:
I love it